### ISSUE TITLE
Critical: Loading a "Game Result" save file in-game causes crashes and loss of progress

### ISSUE THREAD

**Issue by Polymega** [elishacloud/Silent-Hill-2-Enhancements#47]
If you load a "Game Result" save file from the main menu you're greeted with the results/stats screen seen here: 

![image](https://user-images.githubusercontent.com/27395260/45640920-f3491280-ba81-11e8-96fb-6a787b5dfcb5.png)

Then, if you press any button you're brought back to the main menu.

However, if you load a Game Result save file _**while in-game**_ it will either load incorrect game variables, outright soft lock/freeze the game, and [erase current New Game + progression status](https://github.com/elishacloud/Silent-Hill-2-Enhancements/issues/47#issuecomment-507814560). 

----------------------------

### If you load a Game Result file in-game:

| Last Ending You Received: | Behavior: |
| --- | --- |
| Leave | Loads an invisible James high above the ground in the cemetery then crashes the game. ([Picture](https://user-images.githubusercontent.com/27395260/60471855-69df8700-9c34-11e9-951f-a76f53d6e128.png)) |
| Maria | Loads you on the Observation Deck with all items you obtained before the ending. No fog on the Observation Deck. ([Picture](https://user-images.githubusercontent.com/27395260/60471968-eb371980-9c34-11e9-985b-ec1e5d105077.png)) No BGM audio plays. Lighting conditions are completely wrong in the bathroom. ([Picture](https://user-images.githubusercontent.com/27395260/60471973-f5591800-9c34-11e9-82e0-6293d75d317a.png)) |
| In Water | Loads James into a black "void." ([Picture](https://user-images.githubusercontent.com/27395260/60472063-5d0f6300-9c35-11e9-9975-e6b7f1e94211.png)) You cannot control James nor bring up any menus. You must force close the game through Windows. |
| Rebirth | Loads a bugged camera angle of the "Rebirth Island." ([Picture](https://user-images.githubusercontent.com/27395260/60472177-c1cabd80-9c35-11e9-90f8-f5beb6e5e27d.png))  You cannot control James nor bring up any menus. You must force close the game through Windows. |
| DOG | Crashes the game. |
| UFO | Loads you back into Hotel Room 312 (the blue gem is also removed from your inventory). |
| BFaW | Loads the final cutscene of the Born From a Wish ending (Maria throwing her gun over the wall. |

-------------------------------

@Bigmanjapan suggested either locking out the ability to load Game Result specific save files while in-game or, if someone loads a Game Result file while in-game, make it fully follow the steps/process of doing so while on the main menu, instead.

**Comment by Polymega** [comment]
@Bigmanjapan 
Observational note: Only when you load a Game Result save file does the Room ID change to FFFFFFFF. All other times (when you load normal save files) the Room ID changes to 0000000 before loading in the new room. 

When you load a Game Result while in-game it still loads FFFFFFFF first. Perhaps we can use that load call as a trigger to detect if/when someone loads a Game Result file for a possible fix in the future?

**Comment by Polymega** [comment]
The following below may mean nothing, but for these tests I loaded the same Quick Save file and the same Game Result file each time.

Looking at what functions use the Room ID address (01FB7DAC [4 bytes]) the following occurs: 

### If you load a regular save file from the main menu:

![image](https://user-images.githubusercontent.com/27395260/57323356-a4fd8800-70d3-11e9-9f8b-46b3f008694b.png)

```
0053AAA8 - C7 05 AC7DFB01 FFFFFFFF - mov [01FB7DAC],FFFFFFFF
---------------------------------------------------------------------------
0053AAA0 - 8A 45 28  - mov al,[ebp+28]
0053AAA3 - A2 1D81FB01 - mov [sh2pc.exe+1BB811D],al
0053AAA8 - C7 05 AC7DFB01 FFFFFFFF - mov [sh2pc.exe+1BB7DAC],FFFFFFFF <<
0053AAB2 - 8B 4D 00  - mov ecx,[ebp+00]
0053AAB5 - 89 0D E080FB01  - mov [sh2pc.exe+1BB80E0],ecx

EAX=FFFFFFC8
EBX=00000002
ECX=41A00000
EDX=00000000
ESI=01F5CC9C
EDI=01F7A81C
ESP=0018FE58
EBP=01F5CC9C
EIP=0053AAB2
```

### If you load a Game Result save file from the main menu (correct):

![image](https://user-images.githubusercontent.com/27395260/57323400-bcd50c00-70d3-11e9-81c0-2adbf71e8ab1.png)

```
0053AAA8 - C7 05 AC7DFB01 FFFFFFFF - mov [01FB7DAC],FFFFFFFF
---------------------------------------------------------------------------
0053AAA0 - 8A 45 28  - mov al,[ebp+28]
0053AAA3 - A2 1D81FB01 - mov [sh2pc.exe+1BB811D],al
0053AAA8 - C7 05 AC7DFB01 FFFFFFFF - mov [sh2pc.exe+1BB7DAC],FFFFFFFF <<
0053AAB2 - 8B 4D 00  - mov ecx,[ebp+00]
0053AAB5 - 89 0D E080FB01  - mov [sh2pc.exe+1BB80E0],ecx

EAX=000000C8
EBX=0000000F
ECX=41A00000
EDX=00000000
ESI=01F5CC9C
EDI=01F7A81C
ESP=0018FE58
EBP=01F5CC9C
EIP=0053AAB2
```

### If you load a Game Result save file while in-game (incorrect):

![image](https://user-images.githubusercontent.com/27395260/57323781-6caa7980-70d4-11e9-8f81-0050569f0da1.png)

```
0053AAA8 - C7 05 AC7DFB01 FFFFFFFF - mov [01FB7DAC],FFFFFFFF
---------------------------------------------------------------------------
0053AAA0 - 8A 45 28  - mov al,[ebp+28]
0053AAA3 - A2 1D81FB01 - mov [sh2pc.exe+1BB811D],al
0053AAA8 - C7 05 AC7DFB01 FFFFFFFF - mov [sh2pc.exe+1BB7DAC],FFFFFFFF <<
0053AAB2 - 8B 4D 00  - mov ecx,[ebp+00]
0053AAB5 - 89 0D E080FB01  - mov [sh2pc.exe+1BB80E0],ecx

EAX=000000C8
EBX=0000000F
ECX=41A00000
EDX=00000000
ESI=01F5CC9C
EDI=01F7A81C
ESP=0018FE68
EBP=01F5CC9C
EIP=0053AAB2
```

-----------------------

The EAX and EBX are different on the first two. The ESP is different on the last two.

**Comment by Polymega** [comment]
Setting `00932178 byte` to different values invokes different commands in-game:

1 = reload/re-enter into current room
4 = in-game
5 = pull up map
6 = pull up inventory
7 = pull up options menu
8 = pull up bugged Xbox button config menu
9 = pull up save menu
**10/11 = pull up Game Results file**
12 = "COMING SOON!" splash screen
16 = pull up PC pause menu

**Comment by Polymega** [comment]
@Belek666
May I ask for any help/insight you may have on this issue as well? This issue has been stumping us for nearly a year now. It would be very nice to properly fix this as **it can be a game-breaking bug** (more info on that below). 

It's almost as if there's a flag/condition for whenever a save file is loaded from the main menu vs. when it's loaded in-game. Without that flag/condition, loading a Game Result file behaves incorrectly. 

### If you load a Game Result file in-game:

| Last Ending You Received: | Behavior: |
| --- | --- |
| Leave | Loads an invisible James high above the ground in the cemetery then crashes the game. ([Picture](https://user-images.githubusercontent.com/27395260/60471855-69df8700-9c34-11e9-951f-a76f53d6e128.png)) |
| Maria | Loads you on the Observation Deck with all items you obtained before the ending. No fog on the Observation Deck. ([Picture](https://user-images.githubusercontent.com/27395260/60471968-eb371980-9c34-11e9-985b-ec1e5d105077.png)) No BGM audio plays. Lighting conditions are completely wrong in the bathroom. ([Picture](https://user-images.githubusercontent.com/27395260/60471973-f5591800-9c34-11e9-82e0-6293d75d317a.png)) |
| In Water | Loads James into a black "void." ([Picture](https://user-images.githubusercontent.com/27395260/60472063-5d0f6300-9c35-11e9-9975-e6b7f1e94211.png)) You cannot control James nor bring up any menus. You must force close the game through Windows. |
| Rebirth | Loads a bugged camera angel of the "Rebirth Island." ([Picture](https://user-images.githubusercontent.com/27395260/60472177-c1cabd80-9c35-11e9-90f8-f5beb6e5e27d.png))  You cannot control James nor bring up any menus. You must force close the game through Windows. |
| DOG | Crashes the game. |
| UFO | Loads you back into Hotel Room 312 (the blue gem is also removed from your inventory). |

**Comment by Bigmanjapan** [comment]
A major issue with loading Game Result files during gameplay is that they override game statistics of a previous savefile. For example, if a player has 3/4 endings on a particular savefile and wants to grind out a 4/4 ending stat, then an accidental load (during gameplay) of a Game Result file which has a statistic of 1/4 will ruin their progress in a current gameplay session.

If you are to allow players to load up Game Result files during gameplay for them only to see stats on that particular file (if they ever want to do that for some reason) and then be let to continue playing their current walkthrough, then you need to make sure that their previous statistics won't be overriden by that Game Result file.

I still think that disabling an ability to load up Game Result files during gameplay would be a better solution. 

If you take a `Folder 01` which contains only one Game Result file, then one out of these three:

`SaveImage.pcv`
`sh2pc.sys`
`sh2pcsave01.dat`

will contain a byte/bytes that point game engine to `m_card_x_msg_e.mes` file to load a "Game Result" string for it to be displayed in the save/load menu. 

If those bytes could be found, then (I guess) save files could be scanned (upon save/load menu opening, for example), their indexes/number identified and then, the yellow selector could be corrupted once it hits the particular indexes/numbers of a Game Result files to prevent it from working/loading the file.

sh2pc.exe+53C104 (0093C104) — Selected save file index (while in the save/load menu), which is shown as "Data 1", "Data 2" etc

**Comment by Polymega** [comment]
> For example, if a player has 3/4 endings on a particular savefile and wants to grind out a 4/4 ending stat, then an accidental load (during gameplay) of a Game Result file which has a statistic of 1/4 will ruin their progress in a current gameplay session.

Oh wow. Yes... that is extremely bad. Once again, I'm very grateful you've discovered these types of fringe effects for odd game behavior. 

> If those bytes could be found, then (I guess) save files could be scanned (upon save/load menu opening, for example), their indexes/number identified and then, the yellow selector could be corrupted once it hits the particular indexes/numbers of a Game Result files to prevent it from working/loading the file.

Here are SH2 PC Game Result files of all six endings. There are only the six Game Results files in this folder: [Folder 01.zip](https://github.com/elishacloud/Silent-Hill-2-Enhancements/files/3352168/Folder.01.zip)

> I still think that disabling an ability to load up Game Result files during gameplay would be a better solution.

Now knowing everything you said about this, I would agree. It would be detrimental if a player wipes all their progress due to checking a Game Result file in-game (whether on purpose or by accident). 

Assuming we can determine which save files are the Game Result files and can prevent them from being selected on the "Load Game" screen while in-game, do you guys think it would also be a good idea to have the wording grayed-out in the "Load Game" screen while in-game, if it's possible? A quick mock-up: 

![image](https://user-images.githubusercontent.com/27395260/60542616-5854b880-9ce2-11e9-8fc9-f11a2067fe4c.png)

The Game Result text could be yellow like normal in the "Save Game" screen and the player could overwrite the Game Result files in the "Save Game" screen as normal, either in-game or on the main menu.

**Comment by Polymega** [comment]
@Bigmanjapan 
Hypothetically-speaking, let's say we fixed this to where when a user loads a Game Result file while in-game it would take them directly to the Game Results screen like normal. Then afterwards, it brings them back to the main menu. 

If it were done that way, do you think it would still overwrite the current game statistics? Even if it kicked them back to the main menu afterwards and they would have to re-load a previous save point to continue playing like normal? 

No progress has been done for this, but I'm thinking out loud about how an approach _could_ be done for it in the future.

**Comment by Polymega** [comment]
Changing `sh2pc.exe+67A3 (004067A3) mov [00932174],00000016` will change where you go if you load another save file while in-game (instead of the main menu). 

For example, changing it to `mov [00932174],00000006` will bring you back to the main menu if you load another save file in-game.

Changing it to `mov [00932174],0000000C` will play the intro FMV if you load another save file in-game.

The idea of what I was messing around with is to make the game go _back_ to the main menu _before_ loading the new save file. Perhaps if we can divert the instruction process to store `00000016` but have it go to `00000006` (main menu) first it might theoretically fix this critical Game Result bug?

There's more at play within this memory region, but it's above my head.

**Comment by Polymega** [comment]
@elishacloud 
The following is research notes and a work-in-progress pseudo code for a proposed fix to this issue. We will need your input/help to flesh this idea out, please.

There are two ideas/approaches to fix the issue:
1. If a player loads a Game Result file while in-game, show the Game Results stats screen then return the player to the main menu.
2. Prevent the player from selecting/loading a Game Result file while in-game through the Load Screen.

Between the two, Aero_'s preliminary research found a potential way to resolve this using the first method. 

Setting `00932034 byte` to a value of 0 before loading a Game Result file while in-game fixes the issue. However, if you load a non-Game Result file (aka a regular save file) in-game while this value is set to 0 it will crash the game.

So to thought is: While in-game, what if we can detect when the player has a Game Result file highlighted from the Load Screen and set `00932034` to 0? And if/when they scroll to a non-Game Result file while in the Load Screen, set `00932034` back to 1?

![Screenshot 2021-03-20 13-43-25](https://user-images.githubusercontent.com/27395260/111880476-41f10380-8982-11eb-8c54-f81244c79315.png)

The pseudo code (WIP and incomplete) would be something like:

- If `[address needed] (currently in-game or not on main menu)` && [`EventIndex == 0 (Load Screen)`](https://github.com/elishacloud/Silent-Hill-2-Enhancements/blob/master/Patches/Patches.h#L159-L172)
   - If `[code needed] (a Game Result file is currently highlighted)`
      - Set `00932034 == 0` _one write, no constant iteration_
   - ElseIf `[code needed] (a Game Result file is NOT currently highlighted)`
       - Set `00932034 == 1` _one write, no constant iteration_
- ElseIf `[address needed] (currently in-game or not on main menu)` && [`EventIndex != 0 (NOT in Load Screen)`]
   - Set `00932034 == 0`

The "unknown" here--which we'll likely need your help with--is knowing when a user currently has a Game Result file highlighted within the Load Screen menu list, but before they actually select (to load) it. To know this, we'd need to identify a Game Result save file by tracing the code that populates text to the Load/Save Screen. The actual "Game Result" text is stored in `data\etc\message\m_card_x_msg_*.mes` files.

Using pointers, this text string should theoretically be able to be identified. If such a pointer is identified, then we should be able to identify the places in the Load Screen list where Game Result files are located.

Trace the code during the Load Screen menu list population, make a memory buffer that stores info about the order/position of any Game Result file(s) in the player's overall save file list, and use that made up buffer to cross-reference the game's address for the save file list count to identify when a Game Result file is currently highlighted.

The game engine's own memory has a simple selector counter that goes from 0 to whatever number of saves there are in the Load/Save Screen menus. So, tying our own memory buffer position to the game's own selector positions should make this work.

`0093C104 byte` is the address that stores the position of the currently highlighted save file.

Beyond this, I will leave it up to Bigmanjapan to share any further notes/research he deems relevant to the task at hand.

### Other considerations
- A player can delete Game Result files from the Save Screen (different from the Load Screen). They can also save regular save files which can alter the count/position of pre-existing Game Result files. So a memory buffer made on our end would need to be refreshed/re-checked every time the player pulls up the Load Screen while in-game.
- The only issue we know about so far with this approach is that the current BGM audio does not stop when you load a Game Result file while in-game. Ideally, we'd also want to stop the current BGM tracks playing, and ensure the Game Result BGM can still load and play after it is selected from the Load Screen.

**Comment by Bigmanjapan** [comment]
I did some haphazard digging, therefore the information below is quite random.

m_card_x_msg_*.mes files' headers hold offsets for the text strings that come after. Those offsets must be multiplied by 2 to get a true offset (this multiplication is usually done by instructions that poll the .mes files). The offset/pointer for "Game Result" string is 0x05E9. When .mes file is loaded by the game engine, that offset is stored statically at sh2pc.exe+1B52782 (01F52782) 2 bytes address. From what I understand there is one function that uses this address to fetch strings to the code around 0x0044FC9D - 0x0047EEF7 region.

I assume that save/load menu uses sh2pc.sys data (not individual savefile like sh2pcsave00.dat) to determine what text strings to use for each savefile. If this file is cracked, then the issue of knowing where Game Result savefiles are will be solved with ease.

Another way to determine positions of Game Result savefiles might come from reading memory region starting from 0x00933628:

![image](https://user-images.githubusercontent.com/31383389/111881866-2703ab00-89e5-11eb-9a21-b6faa57f3ad6.png)

You can see that every 24 bytes there is a repeating pattern:

Address ---- Value
0x00933628  |  00 00 00 D8
0x00933640  |  00 01 00 D4
0x00933658  |  00 02 FF 00
0x00933670  |  00 03 FF 00

Some savefiles data. Second byte is an index number and if third byte equals to 0xFF, then it's a Game Result file. In this case savefiles 3 and 4 (0x00 = 1) are Game Result savefiles. This region could be scanned for 0xFF values and then indexes could be used to compare between that memory region and currently hightlighted savefile to determine if EE code should be executed.

**Comment by Polymega** [comment]
@elishacloud 
The following is research notes and a work-in-progress pseudo code for a proposed fix to this issue. We will need your input/help to flesh this idea out, please.

There are two ideas/approaches to fix the issue:
1. If a player loads a Game Result file while in-game, show the Game Results stats screen then return the player to the main menu.
2. Prevent the player from selecting/loading a Game Result file while in-game through the Load Screen.

Between the two, Aero_'s preliminary research found a potential way to resolve this using the first method. 

Setting `00932034 byte` to a value of 0 before loading a Game Result file while in-game fixes the issue. However, if you load a non-Game Result file (aka a regular save file) in-game while this value is set to 0 it will crash the game.

So the thought is: While in-game, what if we can detect when the player has a Game Result file highlighted from the Load Screen and set `00932034` to 0? And if/when they scroll to a non-Game Result file while in the Load Screen, set `00932034` back to 1?

![Screenshot 2021-03-20 13-43-25](https://user-images.githubusercontent.com/27395260/111880476-41f10380-8982-11eb-8c54-f81244c79315.png)

The pseudo code (WIP and incomplete) would be something like:

- If [`EventIndex == 0 (Load Screen)`](https://github.com/elishacloud/Silent-Hill-2-Enhancements/blob/master/Patches/Patches.h#L159-L172) && `01DBBF94 float (in-game timer) > 0.0f` _// In the Load Screen while in-game_
   - If `[code needed] (a Game Result file is currently highlighted)`
      - Set `00932034 == 0` _one write, no constant iteration_
   - ElseIf `[code needed] (a Game Result file is NOT currently highlighted)`
       - Set `00932034 == 1` _one write, no constant iteration_
- ElseIf `EventIndex != 0 (NOT in Load Screen)` && `01DBBF94 float (in-game timer) > 0.0f` _// No longer in the Load Screen while in-game_
   - Set `00932034 == 0` _one write, no constant iteration_

---------------------------

- `00932034  byte` is a new address for the project. This is the address that makes loading a Game Result file while in-game work.
- `01DBBF94 float (in-game timer)` is a new address for the project. We use this to determine if the player is currently in-game (if the value is great than 0, we know they're in-game).

The "unknown" here--which we'll likely need your help with--is knowing when a user currently has a Game Result file highlighted within the Load Screen menu list, but before they actually select (to load) it. To know this, we'd need to identify a Game Result save file by tracing the code that populates text to the Load/Save Screen. The actual "Game Result" text is stored in `data\etc\message\m_card_x_msg_*.mes` files.

Using pointers, this text string should theoretically be able to be identified. If such a pointer is identified, then we should be able to identify the places in the Load Screen list where Game Result files are located.

Trace the code during the Load Screen menu list population, make a memory buffer that stores info about the order/position of any Game Result file(s) in the player's overall save file list, and use that made up buffer to cross-reference the game's address for the save file list count to identify when a Game Result file is currently highlighted.

The game engine's own memory has a simple selector counter that goes from 0 to whatever number of saves there are in the Load/Save Screen menus. So, tying our own memory buffer position to the game's own selector positions should make this work.

`0093C104 byte` is the address that stores the position of the currently highlighted save file.

Beyond this, I will leave it up to Bigmanjapan to share any further notes/research he deems relevant to the task at hand.

### Other considerations
- A player can delete Game Result files from the Save Screen (different from the Load Screen). They can also save regular save files which can alter the count/position of pre-existing Game Result files. So a memory buffer made on our end would need to be refreshed/re-checked every time the player pulls up the Load Screen while in-game.
- The only issue we know about so far with this approach is that the current BGM audio does not stop when you load a Game Result file while in-game. Ideally, we'd also want to stop the current BGM tracks playing, and ensure the Game Result BGM can still load and play after it is selected from the Load Screen.

**Comment by Polymega** [comment]
This may not work. If `00932034` is set to 0 and you back out of the Load Screen while in-game, it'll return you to the main menu instead of your current play session. 

So if we change this value to 0 when the player highlights a Game Result save file, and they back out of the Load Screen while a Game Result file is highlighted, it'll incorrectly return them to the main menu instead.

**Comment by elishacloud** [comment]
> Edit: I suppose another approach would be to wait until _after_ the user selects a Game Result file (to load it) and then change `00932034` to 0. But does this address need to be set to 0 _before_ the Game Result file is selected to be loaded in?

This seems like the best idea to me.  Once the user selects the item to be loaded we should then check if it is loading a Game Result entry and if so then change `00932034` to 0.

Question: do game result saves always display in yellow with the text `Game Result`?  If so it would seem like the game already knows which entries are game results and which are not.  Couldn't we just leverage this when deciding when to change `00932034` to 0?

**Comment by Polymega** [comment]
> This seems like the best idea to me. Once the user selects the item to be loaded we should then check if it is loading a Game Result entry and if so then change 00932034 to 0.

Nice. We can try this but note that we don't have any addresses scoped out to know when a user selects a save file. This would be something we'd need to discover.

> Question: do game result saves always display in yellow with the text Game Result? If so it would seem like the game already knows which entries are game results and which are not. Couldn't we just leverage this when deciding when to change 00932034 to 0?

They do always display in yellow, but other save file types also display in yellow (such as Quick Save). It would be better to determine if the save file selected is a Game Result through a method such as how BMJ described.

**Comment by elishacloud** [comment]
It looks like the index for the save entry that you chose is located at `0x009335E4` (byte).

Using this we can have a function that looks like this:

* Get `IndexValue` = value at `0x009335E4` (byte)
* If (`IndexValue` * 24 + 0x933628 + 2) (byte) == 0x50 or 0xFF
  * Set value at `0x00932034` (dword) = 0x00  _one write, no constant iteration_

---------------------------

Here is a proof of concept build (a very quick hack) that I think should fix this:  [d3d8.zip](https://github.com/elishacloud/Silent-Hill-2-Enhancements/files/6216970/d3d8.zip)

_Note: this build only works on binary v1.0_

**Comment by Polymega** [comment]
Well I'll be... This quick proof of concept looks promising! 

So that leaves us with looking into a way of killing the current BGM/SFX playing on a Game Result load while in-game, and also seeing if we can call/play the Game Result BGM.

And I suppose after that it boils down to a lot of playtesting.

**Comment by Polymega** [comment]
Good news! I had a little gremlin in the back of my mind about this, and my memory is a bit fuzzy on its behavior, so I booted up both my PS2 and Xbox console and, when loading a Game Result file from the main menu, **_no music plays._**

So we don't need to worry about playing the Game Result BGM when loading it while in-game. All we'd need to worry about is killing any current audio playing. I'm hoping this would be relatively straightforward though a dsound call similar as you've done to fix the popping/clicking, Elisha?

**Comment by elishacloud** [comment]
> All we'd need to worry about is killing any current audio playing. I'm hoping this would be a relatively straightforward thing to do, similar to how you fixed the popping/clicking, Elisha?

Yes, it is possible.  However, this just cuts off the sound as soon as the game enters the Game Results page.  Ideally we would find out what address or code tells the game to stop playing music (like what happens when you go to the main menu) and trigger that.

Here is a testing build.  It should have everything addressed that we have discussed so far: [d3d8.zip](https://github.com/elishacloud/Silent-Hill-2-Enhancements/files/6218767/d3d8.zip)

**Comment by Polymega** [comment]
> Ideally we would find out what address or code tells the game to stop playing music (like what happens when you go to the main menu) and trigger that.

@Bigmanjapan would you be able to help locate this? I tried searching for the instruction responsible for this a few months back and admitted defeat after many hours of unsuccessfully poking and prodding. (I had no clue for a reference/starting point to look.)

> Yes, it is possible. However, this just cuts off the sound as soon as the game enters the Game Results page.

Thanks Elisha. We actually want the game to stop all current sound when loading a Game Result file, so that's acting correct with this new test build. 

Are you meaning it also cuts off the "game load" SFX early? If so, the game already cuts this SFX early when loading _any_ save file while in-game. [It would be nice to prevent that from happening.](https://github.com/elishacloud/Silent-Hill-2-Enhancements/issues/209)

Although, with this test build, the "game load" SFX also gets cut off when loading a Game Result through the main menu (where that shouldn't happen). A way to prevent this would be adding a conditional check to see if the file is being loaded while in-game. That way, the "game load" SFX only gets cut off early when loading it while in-game (like any other save file). Something like:

- Get `IndexValue` = value at `0x009335E4` (byte)
- If value at (`IndexValue` * 24 + 0x933628 + 2) (byte) == 0x50 or 0xFF && `01DBBF94 float (in-game timer)` > 0.0f
   - Set value at `0x00932034` (dword) = 0x00 _one write, no constant iteration_
   - Stop current audio playing

... I'm not sure if we'll ever find a solution to preserve the "game load" SFX when loading save files while in-game, but that conditional check above would at least put its audio behavior in line with loading any other save file.

**Comment by elishacloud** [comment]
> Are you meaning it also cuts off the "game load" SFX early? If so, the game already cuts this SFX early when loading _any_ save file while in-game.

Ok, then it is probably ok for now.  This just works the same as everything else.  We can fix this when we fix #209.

> Although, with this test build, the "game load" SFX also gets cut off when loading a Game Result through the main menu (where that shouldn't happen). A way to prevent this would be [...] Something like:

I changed the logic a bit.  I always want to set the value at `0x009335E4` (byte) but only stop the audio when the game is playing.  Also, I am using the RoomID to determine whether you are in game or not.  The RoomID seems to always be set no mater where you are in the game and it is never set when in the main menu.

I am doing something like this:

 * Get `IndexValue` = value at `0x009335E4` (byte)
 * If value at (`IndexValue` * 24 + 0x933628 + 2) (byte) == 0x50 or 0xFF
   * Set value at `0x00932034` (dword) = 0x00 _one write, no constant iteration_
   * If  RoomID is not 0
     * Stop current audio playing

Here is the new test build:  [d3d8.zip](https://github.com/elishacloud/Silent-Hill-2-Enhancements/files/6219139/d3d8.zip)

**Comment by Bigmanjapan** [comment]
> Ideally we would find out what address or code tells the game to stop playing music (like what happens when you go to the main menu) and trigger that.

I loaded a location with active background music, then set sh2pc.exe+532174 (00932174) to 0x00 to initialize main menu loading process. The result is main menu on screen with background music being active.

I found this address to stop background music when it's value is set to 0x00. For some tracks only single lower or higher byte is changed, therefore be sure to null all 4 bytes.

sh2pc.exe+197EAE0 (01DEAE0)
